File: PULSE.Services/Data/Products/Bicycle.cs
namespace PULSE.Services.Data
{
    public class Bicycle : Product
    {
        public Bicycle()
        {
            AvailableSizes = new HashSet<AvailableSize>();
        }

        public decimal? Weight { get; set; }
        public int? ProductionYear { get; set; }
        public string? WheelSize { get; set; }

        public virtual BicycleCategory? ProductCategory { get; set; }
        public virtual ICollection<AvailableSize> AvailableSizes { get; set; }
    }
}

File: pulse_admin/lib/api/types/bicycle/bicycle.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:pulse_admin/api/types/available_size/available_size.dart';
import 'package:pulse_admin/api/types/product_image/product_image.dart';

import '../brand/brand.dart';
import '../abstract/product.dart';
import '../product_category/product_category.dart';

part 'bicycle.g.dart';

@JsonSerializable()
class Bicycle extends Product {
  static const String discriminator = "Bicycle";

  double? weight;
  int? productionYear;
  String? wheelSize;
  List<AvailableSize>? availableSizes;

  Bicycle();

  factory Bicycle.fromJson(Map<String, dynamic> json) =>
      _$BicycleFromJson(json);

  Map<String, dynamic> toJson() => _$BicycleToJson(this);

  bool isAvailable() {
    return (availableSizes
                ?.map((e) => e.availableQty ?? 0)
                .reduce((a, b) => a + b) ??
            0) >
        0;
  }

  @override
  String getDiscriminator() => discriminator;
}
